Per-message deflate as described in RFC 7692

hoytech opened the pull request:
Github issue #5

For my work-load (somewhat redundant JSON) it reduces the transmitted
message sizes to nearly 1/10th their uncompressed sizes so it will be
a huge bandwidth saver.

It basically does the bare minimum to be RFC 7692 compatible. I have
tested the compression with chromium and websocketpp 0.7.0 and seasocks
seems to compress/decompress messages properly in both directions.

Current limitations:
  - Can't configure the *_no_context_takeover or *_max_window_bits
    extensions. It will only negotiate the default (context takeover
    and max window bits in both directions)
  - All transmitted messages will be compressed. Ideally you should
    be able to opt-out of compressing particular messages (but it
    does support clients that compress particular messages)
  - Only ever issues one DEFLATE block per message since the
    seasocks API doesn't support incremental message construction
    (but I believe it does support receiving multiple blocks from
    clients that do)
  - Various tuning parameters such as buffer size and memLevel
    are hard-coded. Some apps might like to customise them
  - Requires linking to zlib even if no compression will be used
  - No tests :(

offa commented:
You don't need to call `find_package(ZLIB)` every time; just call it once in the Toplevel CMakeLists.txt.

hoytech commented:
@offa - Ahh OK didn't know that thanks. I'm a cmake noob. I tried moving it into the CMakeLists.txt file in the root, but was getting linking errors. I'll try fooling with it a bit more...

offa commented:
The `find_package()` searches for the library and stores into `${ZLIB_LIBRARIES}` (which is finally linked).

If you still get an error, you can post it here, maybe I can help you.

hoytech commented:
OK got it working. I needed to put it above the add_subdirectory() calls which makes sense I guess. Thanks!

mattgodbolt commented:
Nice! This is great! Thank you!

mattgodbolt commented on src/main/c/internal/ZlibContext.h:
```diff
@@ -0,0 +1,103 @@
+
+
+class ZlibContext {
+public:
```
Probably best to make this non-copyable. I'll happily do that in a follow-up change unless you agree and get to it first :)

hoytech replied:
Yes I agree, this should hopefully never be copied.

Also, it just occurred to me that this class is living in a global namespace.. You could wrap that in a seasocks namespace if you like.

mattgodbolt commented on src/main/c/internal/ZlibContext.h:
```diff
@@ -0,0 +1,103 @@
+
+    bool inflate(std::vector<uint8_t>& input, std::vector<uint8_t>& output, int& zlibError) {
+        // Append 4 octets prior to decompression (see RFC 7692, section 7.2.2)
+        uint8_t tail_end[4] = {0x00, 0x00, 0xff, 0xff};
```
I get that you need to put the four bytes on the end; but mutating the input vector is a little surprising. It's cool as is, but I thought I'd make that observation. It would be more programming work to supply the extra four bytes in a separate call to `inflate()`, and/or more CPU work to copy the vector to append four bytes to it...so I get why :)

hoytech replied:
Yes, good point. At the very least it should document what it's doing!

hoytech commented:
I pushed a new version of the diff:

* Rebased onto master
* Deleted copy constructor of ZlibContext
* Put ZlibContext into seasocks namespace
* Documented in comments that inflate() modifies the input param

hoytech commented:
Thanks! Not sure how the approve thing works but it says "Only those with write access to this repository can merge pull requests."

mattgodbolt commented:
Thanks for the changes. I wonder if I need to approve again...I'll do so; let me know ASAP if it doesn't work and I'll merge it myself :)

hoytech commented:
Nope. Or it's possible I'm just dumb and can't see the right button to press :)

mattgodbolt commented:
Not to worry! Just trying to use GH features if they work :) Thanks so much!

hoytech commented:
Yay, thanks!
